File: rust/scene-syncd/src/procedural/sdf.rs
use glam::Vec3;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct SdfBounds {
    pub min: Vec3,
    pub max: Vec3,
}

impl SdfBounds {
    pub fn new(min: Vec3, max: Vec3) -> Self {
        Self { min, max }
    }

    pub fn merge(&self, other: &SdfBounds) -> SdfBounds {
        SdfBounds {
            min: self.min.min(other.min),
            max: self.max.max(other.max),
        }
    }

    pub fn expand(&self, padding: f32) -> SdfBounds {
        let p = Vec3::splat(padding);
        SdfBounds {
            min: self.min - p,
            max: self.max + p,
        }
    }

    pub fn size(&self) -> Vec3 {
        self.max - self.min
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SdfPrimitive {
    Sphere {
        center: [f32; 3],
        radius: f32,
    },
    Box {
        min: [f32; 3],
        max: [f32; 3],
    },
    Capsule {
        start: [f32; 3],
        end: [f32; 3],
        radius: f32,
    },
    Torus {
        center: [f32; 3],
        major_radius: f32,
        minor_radius: f32,
    },
    Gyroid {
        frequency: f32,
        thickness: f32,
    },
    Scherk {
        frequency: f32,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SdfTree {
    Primitive(SdfPrimitive),
    Union(Box<SdfTree>, Box<SdfTree>, f32),
    Difference(Box<SdfTree>, Box<SdfTree>, f32),
    Intersection(Box<SdfTree>, Box<SdfTree>, f32),
    Transform(Box<SdfTree>, [f32; 16]),
    DomainWarp(Box<SdfTree>, f32, f32),
}

impl SdfPrimitive {
    pub fn evaluate(&self, point: Vec3) -> f32 {
        match self {
            SdfPrimitive::Sphere { center, radius } => {
                let c = Vec3::from(*center);
                (point - c).length() - radius
            }
            SdfPrimitive::Box { min, max } => {
                let bmin = Vec3::from(*min);
                let bmax = Vec3::from(*max);
                let center = (bmin + bmax) * 0.5;
                let half = (bmax - bmin) * 0.5;
                let d = (point - center).abs() - half;
                d.max(Vec3::ZERO).length() + d.x.max(d.y).max(d.z).min(0.0)
            }
            SdfPrimitive::Capsule { start, end, radius } => {
                let a = Vec3::from(*start);
                let b = Vec3::from(*end);
                let pa = point - a;
                let ba = b - a;
                let denom = ba.dot(ba);
                let h = if denom > 0.0 {
                    (pa.dot(ba) / denom).clamp(0.0, 1.0)
                } else {
                    0.0
                };
                (pa - ba * h).length() - *radius
            }
            SdfPrimitive::Torus {
                center,
                major_radius,
                minor_radius,
            } => {
                let c = Vec3::from(*center);
                let p = point - c;
                let q = glam::vec2(glam::vec2(p.x, p.y).length() - *major_radius, p.z);
                q.length() - minor_radius
            }
            SdfPrimitive::Gyroid {
                frequency,
                thickness,
            } => {
                let f = *frequency;
                let g = f32::sin(point.x * f) * f32::sin(point.y * f) * f32::cos(point.z * f)
                    + f32::sin(point.z * f) * f32::sin(point.x * f) * f32::cos(point.y * f)
                    + f32::sin(point.y * f) * f32::sin(point.z * f) * f32::cos(point.x * f);
                g.abs() - *thickness
            }
            SdfPrimitive::Scherk { frequency } => {
                let f = *frequency;
                let s = f32::sinh(point.x * f) * f32::sinh(point.y * f) - f32::sin(point.z * f);
                s.abs() - 0.1
            }
        }
    }

    pub fn estimate_bounds(&self) -> Option<SdfBounds> {
        match self {
            SdfPrimitive::Sphere { center, radius } => {
                let c = Vec3::from(*center);
                let r = *radius;
                Some(SdfBounds::new(c - Vec3::splat(r), c + Vec3::splat(r)))
            }
            SdfPrimitive::Box { min, max } => {
                Some(SdfBounds::new(Vec3::from(*min), Vec3::from(*max)))
            }
            SdfPrimitive::Capsule { start, end, radius } => {
                let a = Vec3::from(*start);
                let b = Vec3::from(*end);
                let r = Vec3::splat(*radius);
                Some(SdfBounds::new(a.min(b) - r, a.max(b) + r))
            }
            SdfPrimitive::Torus {
                center,
                major_radius,
                minor_radius,
            } => {
                let c = Vec3::from(*center);
                let outer = major_radius + minor_radius;
                Some(SdfBounds::new(
                    c - Vec3::new(outer, outer, *minor_radius),
                    c + Vec3::new(outer, outer, *minor_radius),
                ))
            }
            SdfPrimitive::Gyroid { .. } | SdfPrimitive::Scherk { .. } => None,
        }
    }
}

impl SdfTree {
    pub fn evaluate(&self, point: Vec3) -> f32 {
        match self {
            SdfTree::Primitive(p) => p.evaluate(point),
            SdfTree::Union(a, b, smoothness) => {
                let da = a.evaluate(point);
                let db = b.evaluate(point);
                if *smoothness <= 0.0 {
                    da.min(db)
                } else {
                    smooth_min(da, db, *smoothness)
                }
            }
            SdfTree::Difference(a, b, smoothness) => {
                let da = a.evaluate(point);
                let db = b.evaluate(point);
                if *smoothness <= 0.0 {
                    (-db).max(da)
                } else {
                    smooth_min(-db, da, *smoothness)
                }
            }
            SdfTree::Intersection(a, b, smoothness) => {
                let da = a.evaluate(point);
                let db = b.evaluate(point);
                if *smoothness <= 0.0 {
                    da.max(db)
                } else {
                    smooth_min(da, db, *smoothness)
                }
            }
            SdfTree::Transform(child, mat) => {
                let m = glam::Mat4::from_cols_array(mat);
                let inv = m.inverse();
                let transformed = inv.transform_point3(point);
                child.evaluate(transformed)
            }
            SdfTree::DomainWarp(child, amplitude, frequency) => {
                let f = *frequency;
                let warp = Vec3::new(
                    f32::sin(point.y * f + 11.17) * f32::cos(point.z * f + 3.31),
                    f32::sin(point.z * f + 17.71) * f32::cos(point.x * f + 5.37),
                    f32::sin(point.x * f + 23.13) * f32::cos(point.y * f + 7.91),
                ) * *amplitude;
                child.evaluate(point + warp)
            }
        }
    }

    pub fn normal(&self, point: Vec3, epsilon: f32) -> Vec3 {
        let ex = self.evaluate(Vec3::new(point.x + epsilon, point.y, point.z))
            - self.evaluate(Vec3::new(point.x - epsilon, point.y, point.z));
        let ey = self.evaluate(Vec3::new(point.x, point.y + epsilon, point.z))
            - self.evaluate(Vec3::new(point.x, point.y - epsilon, point.z));
        let ez = self.evaluate(Vec3::new(point.x, point.y, point.z + epsilon))
            - self.evaluate(Vec3::new(point.x, point.y, point.z - epsilon));
        let n = Vec3::new(ex, ey, ez);
        let len = n.length();
        if len > 0.0 {
            n / len
        } else {
            Vec3::Z
        }
    }

    pub fn estimate_bounds(&self) -> Option<SdfBounds> {
        match self {
            SdfTree::Primitive(p) => p.estimate_bounds(),
            SdfTree::Union(a, b, _) | SdfTree::Intersection(a, b, _) => {
                match (a.estimate_bounds(), b.estimate_bounds()) {
                    (Some(ba), Some(bb)) => Some(ba.merge(&bb)),
                    (Some(ba), None) => Some(ba),
                    (None, Some(bb)) => Some(bb),
                    (None, None) => None,
                }
            }
            SdfTree::Difference(a, b, _) => a.estimate_bounds().or_else(|| b.estimate_bounds()),
            SdfTree::Transform(child, mat) => child.estimate_bounds().map(|bounds| {
                let m = glam::Mat4::from_cols_array(mat);
                let corners = [
                    Vec3::new(bounds.min.x, bounds.min.y, bounds.min.z),
                    Vec3::new(bounds.max.x, bounds.min.y, bounds.min.z),
                    Vec3::new(bounds.min.x, bounds.max.y, bounds.min.z),
                    Vec3::new(bounds.min.x, bounds.min.y, bounds.max.z),
                    Vec3::new(bounds.max.x, bounds.max.y, bounds.min.z),
                    Vec3::new(bounds.max.x, bounds.min.y, bounds.max.z),
                    Vec3::new(bounds.min.x, bounds.max.y, bounds.max.z),
                    Vec3::new(bounds.max.x, bounds.max.y, bounds.max.z),
                ];
                let transformed: Vec<Vec3> =
                    corners.iter().map(|c| m.transform_point3(*c)).collect();
                let min = transformed.iter().fold(Vec3::MAX, |a, &b| a.min(b));
                let max = transformed.iter().fold(Vec3::MIN, |a, &b| a.max(b));
                SdfBounds::new(min, max)
            }),
            SdfTree::DomainWarp(child, amplitude, _) => {
                child.estimate_bounds().map(|bounds| bounds.expand(*amplitude))
            }
        }
    }
}

fn smooth_min(a: f32, b: f32, k: f32) -> f32 {
    let h = (k - (a - b).abs()).max(0.0) / k;
    a.min(b) - h * h * k * 0.25
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sphere_sdf() {
        let sphere = SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        };
        assert!(sphere.evaluate(Vec3::ZERO) < 0.0);
        assert!(sphere.evaluate(Vec3::new(0.0, 0.0, 2.0)) > 0.0);
        let surface = sphere.evaluate(Vec3::new(1.0, 0.0, 0.0));
        assert!(surface.abs() < 1e-4);
    }

    #[test]
    fn test_box_sdf() {
        let box_ = SdfPrimitive::Box {
            min: [-1.0, -1.0, -1.0],
            max: [1.0, 1.0, 1.0],
        };
        assert!(box_.evaluate(Vec3::ZERO) < 0.0);
        assert!(box_.evaluate(Vec3::new(2.0, 0.0, 0.0)) > 0.0);
    }

    #[test]
    fn test_capsule_sdf() {
        let capsule = SdfPrimitive::Capsule {
            start: [0.0, 0.0, 0.0],
            end: [2.0, 0.0, 0.0],
            radius: 0.5,
        };
        assert!(capsule.evaluate(Vec3::new(1.0, 0.0, 0.0)) < 0.0);
        assert!(capsule.evaluate(Vec3::new(1.0, 1.0, 0.0)) > 0.0);
    }

    #[test]
    fn test_union_hard() {
        let a = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        });
        let b = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [3.0, 0.0, 0.0],
            radius: 1.0,
        });
        let union = SdfTree::Union(Box::new(a), Box::new(b), 0.0);
        assert!(union.evaluate(Vec3::ZERO) < 0.0);
        assert!(union.evaluate(Vec3::new(3.0, 0.0, 0.0)) < 0.0);
        assert!(union.evaluate(Vec3::new(1.5, 0.0, 0.0)) > 0.0);
    }

    #[test]
    fn test_union_smooth() {
        let a = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        });
        let b = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [3.0, 0.0, 0.0],
            radius: 1.0,
        });
        let union = SdfTree::Union(Box::new(a), Box::new(b), 0.5);
        let mid = union.evaluate(Vec3::new(1.5, 0.0, 0.0));
        assert!(mid < 1.0);
    }

    #[test]
    fn test_difference() {
        let a = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        });
        let b = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.5, 0.0, 0.0],
            radius: 0.8,
        });
        let diff = SdfTree::Difference(Box::new(a), Box::new(b), 0.0);
        assert!(diff.evaluate(Vec3::new(-0.9, 0.0, 0.0)) < 0.0);
        assert!(diff.evaluate(Vec3::new(0.5, 0.0, 0.0)) > 0.0);
    }

    #[test]
    fn test_intersection() {
        let a = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        });
        let b = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.5, 0.0, 0.0],
            radius: 1.0,
        });
        let isect = SdfTree::Intersection(Box::new(a), Box::new(b), 0.0);
        assert!(isect.evaluate(Vec3::new(0.25, 0.0, 0.0)) < 0.0);
        assert!(isect.evaluate(Vec3::new(2.0, 0.0, 0.0)) > 0.0);
    }

    #[test]
    fn test_normal_accuracy() {
        let sphere = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        });
        let n = sphere.normal(Vec3::new(1.0, 0.0, 0.0), 0.001);
        let expected = Vec3::X;
        let err = (n - expected).length();
        assert!(err < 0.01, "normal error = {err}, n = {n}");
    }

    #[test]
    fn test_estimate_bounds_sphere() {
        let sphere = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [1.0, 2.0, 3.0],
            radius: 2.0,
        });
        let bounds = sphere.estimate_bounds().unwrap();
        assert!((bounds.min.x - (-1.0)).abs() < 1e-4);
        assert!((bounds.max.x - 3.0).abs() < 1e-4);
    }

    #[test]
    fn test_gyroid_no_bounds() {
        let gyroid = SdfTree::Primitive(SdfPrimitive::Gyroid {
            frequency: 1.0,
            thickness: 0.1,
        });
        assert!(gyroid.estimate_bounds().is_none());
    }

    #[test]
    fn test_union_bounds_merge() {
        let a = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        });
        let b = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [5.0, 0.0, 0.0],
            radius: 1.0,
        });
        let union = SdfTree::Union(Box::new(a), Box::new(b), 0.0);
        let bounds = union.estimate_bounds().unwrap();
        assert!(bounds.max.x > 5.0);
    }

    #[test]
    fn test_domain_warp_expands_bounds() {
        let sphere = SdfTree::Primitive(SdfPrimitive::Sphere {
            center: [0.0, 0.0, 0.0],
            radius: 1.0,
        });
        let warped = SdfTree::DomainWarp(Box::new(sphere), 0.5, 1.0);
        let bounds = warped.estimate_bounds().unwrap();
        assert!(bounds.min.x < -1.0);
        assert!(bounds.max.x > 1.0);
    }
}

File: Python/server/generation/sdf.py
"""Pure-Python SDF evaluation and coarse mesh extraction."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence, Tuple

from server.generation.geometry import Aabb3, Vector3, vec3
from server.generation.mesh_payload import ProceduralMeshPayload


# --- 3D Value Noise (hash-based, no external dependencies) ---

def _hash3(ix: int, iy: int, iz: int) -> float:
    """Deterministic pseudo-random float in [-1, 1] from integer coordinates."""
    n = ix * 73856093 ^ iy * 19349663 ^ iz * 83492791
    n = (n ^ (n >> 13)) * 1274126177
    n = n ^ (n >> 16)
    return ((n & 0x7FFFFFFF) / 0x7FFFFFFF) * 2.0 - 1.0


def _lerp(a: float, b: float, t: float) -> float:
    return a + t * (b - a)


def _smoothstep(t: float) -> float:
    return t * t * (3.0 - 2.0 * t)


def value_noise_3d(x: float, y: float, z: float) -> float:
    """3D value noise with trilinear interpolation."""
    ix = math.floor(x)
    iy = math.floor(y)
    iz = math.floor(z)
    fx = _smoothstep(x - ix)
    fy = _smoothstep(y - iy)
    fz = _smoothstep(z - iz)

    c000 = _hash3(ix, iy, iz)
    c100 = _hash3(ix + 1, iy, iz)
    c010 = _hash3(ix, iy + 1, iz)
    c110 = _hash3(ix + 1, iy + 1, iz)
    c001 = _hash3(ix, iy, iz + 1)
    c101 = _hash3(ix + 1, iy, iz + 1)
    c011 = _hash3(ix, iy + 1, iz + 1)
    c111 = _hash3(ix + 1, iy + 1, iz + 1)

    return _lerp(
        _lerp(_lerp(c000, c100, fx), _lerp(c010, c110, fx), fy),
        _lerp(_lerp(c001, c101, fx), _lerp(c011, c111, fx), fy),
        fz,
    )


def fbm_3d(x: float, y: float, z: float, octaves: int = 6, lacunarity: float = 2.0, gain: float = 0.5) -> float:
    """Fractal Brownian motion using 3D value noise."""
    value = 0.0
    amplitude = 1.0
    frequency = 1.0
    norm = 0.0
    for _ in range(max(1, min(octaves, 10))):
        value += amplitude * value_noise_3d(x * frequency, y * frequency, z * frequency)
        norm += amplitude
        amplitude *= gain
        frequency *= lacunarity
    return value / max(norm, 1e-6)


def _sub(a: Vector3, b: Vector3) -> Vector3:
    return (a[0] - b[0], a[1] - b[1], a[2] - b[2])


def _add(a: Vector3, b: Vector3) -> Vector3:
    return (a[0] + b[0], a[1] + b[1], a[2] + b[2])


def _mul(a: Vector3, s: float) -> Vector3:
    return (a[0] * s, a[1] * s, a[2] * s)


def _dot(a: Vector3, b: Vector3) -> float:
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]


def _length(a: Vector3) -> float:
    return math.sqrt(max(_dot(a, a), 0.0))


def _normalize(a: Vector3, fallback: Vector3 = (0.0, 0.0, 1.0)) -> Vector3:
    length = _length(a)
    if length <= 1.0e-9:
        return fallback
    return (a[0] / length, a[1] / length, a[2] / length)


def smooth_min(a: float, b: float, k: float) -> float:
    if k <= 0.0:
        return min(a, b)
    h = max(k - abs(a - b), 0.0) / k
    return min(a, b) - h * h * k * 0.25


def evaluate_sdf(node: Mapping[str, Any], point: Sequence[float]) -> float:
    """Evaluate Rust-compatible SDF JSON at ``point``."""
    p = vec3(point)
    node_type = str(node.get("type", "sphere")).lower()
    if node_type in {"primitive"} and isinstance(node.get("primitive"), Mapping):
        return evaluate_sdf(node["primitive"], p)
    if node_type == "sphere":
        center = vec3(node.get("center", (0.0, 0.0, 0.0)))
        return _length(_sub(p, center)) - float(node.get("radius", 100.0))
    if node_type == "box":
        bmin = vec3(node.get("min", (-100.0, -100.0, -100.0)))
        bmax = vec3(node.get("max", (100.0, 100.0, 100.0)))
        center = _mul(_add(bmin, bmax), 0.5)
        half = _mul(_sub(bmax, bmin), 0.5)
        d = (abs(p[0] - center[0]) - half[0], abs(p[1] - center[1]) - half[1], abs(p[2] - center[2]) - half[2])
        outside = _length((max(d[0], 0.0), max(d[1], 0.0), max(d[2], 0.0)))
        inside = min(max(d[0], max(d[1], d[2])), 0.0)
        return outside + inside
    if node_type == "capsule":
        start = vec3(node.get("start", (0.0, 0.0, 0.0)))
        end = vec3(node.get("end", (100.0, 0.0, 0.0)))
        radius = float(node.get("radius", 30.0))
        pa = _sub(p, start)
        ba = _sub(end, start)
        denom = _dot(ba, ba)
        h = min(max(_dot(pa, ba) / denom, 0.0), 1.0) if denom > 0.0 else 0.0
        return _length(_sub(pa, _mul(ba, h))) - radius
    if node_type == "torus":
        center = vec3(node.get("center", (0.0, 0.0, 0.0)))
        q = _sub(p, center)
        major = float(node.get("major_radius", 100.0))
        minor = float(node.get("minor_radius", 30.0))
        return math.hypot(math.hypot(q[0], q[1]) - major, q[2]) - minor
    if node_type == "gyroid":
        f = float(node.get("frequency", 1.0))
        thickness = float(node.get("thickness", 0.1))
        g = (
            math.sin(p[0] * f) * math.sin(p[1] * f) * math.cos(p[2] * f)
            + math.sin(p[2] * f) * math.sin(p[0] * f) * math.cos(p[1] * f)
            + math.sin(p[1] * f) * math.sin(p[2] * f) * math.cos(p[0] * f)
        )
        return abs(g) - thickness
    if node_type == "scherk":
        f = float(node.get("frequency", 1.0))
        return abs(math.sinh(p[0] * f) * math.sinh(p[1] * f) - math.sin(p[2] * f)) - 0.1
    if node_type == "union":
        children = [child for child in node.get("children", []) if isinstance(child, Mapping)]
        if not children:
            return 1.0e9
        k = float(node.get("smoothness", 0.0))
        value = evaluate_sdf(children[0], p)
        for child in children[1:]:
            value = smooth_min(value, evaluate_sdf(child, p), k)
        return value
    if node_type == "difference":
        a = node.get("a") or node.get("left") or node.get("child")
        b = node.get("b") or node.get("right") or node.get("subtract")
        if isinstance(a, Mapping) and isinstance(b, Mapping):
            return max(evaluate_sdf(a, p), -evaluate_sdf(b, p))
        return 1.0e9
    if node_type == "intersection":
        children = [child for child in node.get("children", []) if isinstance(child, Mapping)]
        if not children:
            return 1.0e9
        return max(evaluate_sdf(child, p) for child in children)
    if node_type == "domain_warp":
        child = node.get("child")
        if not isinstance(child, Mapping):
            return 1.0e9
        amplitude = float(node.get("amplitude", 0.0))
        frequency = float(node.get("frequency", 0.01))
        warp = (
            math.sin(p[1] * frequency + 11.17) * math.cos(p[2] * frequency + 3.31) * amplitude,
            math.sin(p[2] * frequency + 17.71) * math.cos(p[0] * frequency + 5.37) * amplitude,
            math.sin(p[0] * frequency + 23.13) * math.cos(p[1] * frequency + 7.91) * amplitude,
        )
        return evaluate_sdf(child, _add(p, warp))
    if node_type == "displace":
        child = node.get("child")
        if not isinstance(child, Mapping):
            return 1.0e9
        amp = float(node.get("amplitude", 10.0))
        freq = float(node.get("frequency", 0.01))
        octaves = int(node.get("octaves", 4))
        decay = float(node.get("amplitude_decay", 0.5))
        noise = fbm_3d(p[0] * freq, p[1] * freq, p[2] * freq, octaves=octaves, gain=decay)
        return evaluate_sdf(child, p) + amp * noise
    return 1.0e9


def estimate_sdf_bounds(node: Mapping[str, Any], default: Optional[Aabb3] = None) -> Aabb3:
    node_type = str(node.get("type", "")).lower()
    if node_type == "sphere":
        center = vec3(node.get("center", (0.0, 0.0, 0.0)))
        radius = float(node.get("radius", 100.0))
        return Aabb3((center[0] - radius, center[1] - radius, center[2] - radius), (center[0] + radius, center[1] + radius, center[2] + radius))
    if node_type == "box":
        return Aabb3(vec3(node.get("min", (-100.0, -100.0, -100.0))), vec3(node.get("max", (100.0, 100.0, 100.0))))
    if node_type == "capsule":
        start = vec3(node.get("start", (0.0, 0.0, 0.0)))
        end = vec3(node.get("end", (100.0, 0.0, 0.0)))
        radius = float(node.get("radius", 30.0))
        return Aabb3(
            (min(start[0], end[0]) - radius, min(start[1], end[1]) - radius, min(start[2], end[2]) - radius),
            (max(start[0], end[0]) + radius, max(start[1], end[1]) + radius, max(start[2], end[2]) + radius),
        )
    children = []
    if isinstance(node.get("child"), Mapping):
        children.append(node["child"])
    children.extend(child for child in node.get("children", []) if isinstance(child, Mapping))
    bounds = [estimate_sdf_bounds(child, default) for child in children]
    if bounds:
        merged = bounds[0]
        for item in bounds[1:]:
            merged = merged.merge(item)
        if node_type == "domain_warp":
            return merged.expand(float(node.get("amplitude", 0.0)))
        if node_type == "displace":
            return merged.expand(abs(float(node.get("amplitude", 0.0))))
        return merged
    return default or Aabb3((-500.0, -500.0, -500.0), (500.0, 500.0, 500.0))


def sdf_normal(node: Mapping[str, Any], point: Sequence[float], epsilon: float) -> Vector3:
    p = vec3(point)
    ex = evaluate_sdf(node, (p[0] + epsilon, p[1], p[2])) - evaluate_sdf(node, (p[0] - epsilon, p[1], p[2]))
    ey = evaluate_sdf(node, (p[0], p[1] + epsilon, p[2])) - evaluate_sdf(node, (p[0], p[1] - epsilon, p[2]))
    ez = evaluate_sdf(node, (p[0], p[1], p[2] + epsilon)) - evaluate_sdf(node, (p[0], p[1], p[2] - epsilon))
    return _normalize((ex, ey, ez))


def sdf_to_voxel_surface_mesh(
    sdf_tree: Mapping[str, Any],
    bounds: Optional[Aabb3] = None,
    resolution: int = 32,
    mcp_id: str = "sdf_mesh",
    request_id: int = 1,
) -> ProceduralMeshPayload:
    """Extract a coarse manifold-ish voxel surface mesh from an SDF.

    This is intentionally pure Python and deterministic. For production smooth
    meshes, use the UE/scene-syncd marching-cubes path; this path gives Python
    a local generator implementation for validation and migration work.
    """
    res = max(4, min(int(resolution), 96))
    if bounds is None:
        estimated = estimate_sdf_bounds(sdf_tree)
        max_extent = max(estimated.size())
        b = estimated.expand(max(max_extent * 0.08, 0.01))
    else:
        b = bounds
    size = b.size()
    step = (size[0] / res, size[1] / res, size[2] / res)
    inside = set()
    for ix in range(res):
        x = b.min[0] + (ix + 0.5) * step[0]
        for iy in range(res):
            y = b.min[1] + (iy + 0.5) * step[1]
            for iz in range(res):
                z = b.min[2] + (iz + 0.5) * step[2]
                if evaluate_sdf(sdf_tree, (x, y, z)) < 0.0:
                    inside.add((ix, iy, iz))

    positions: List[List[float]] = []
    normals: List[List[float]] = []
    indices: List[int] = []
    faces = [
        ((1, 0, 0), [(1, 0, 0), (1, 1, 0), (1, 1, 1), (1, 0, 1)]),
        ((-1, 0, 0), [(0, 0, 0), (0, 0, 1), (0, 1, 1), (0, 1, 0)]),
        ((0, 1, 0), [(0, 1, 0), (0, 1, 1), (1, 1, 1), (1, 1, 0)]),
        ((0, -1, 0), [(0, 0, 0), (1, 0, 0), (1, 0, 1), (0, 0, 1)]),
        ((0, 0, 1), [(0, 0, 1), (1, 0, 1), (1, 1, 1), (0, 1, 1)]),
        ((0, 0, -1), [(0, 0, 0), (0, 1, 0), (1, 1, 0), (1, 0, 0)]),
    ]
    for cell in inside:
        ix, iy, iz = cell
        base_origin = (b.min[0] + ix * step[0], b.min[1] + iy * step[1], b.min[2] + iz * step[2])
        for direction, corners in faces:
            neighbor = (ix + direction[0], iy + direction[1], iz + direction[2])
            if neighbor in inside:
                continue
            base = len(positions)
            for corner in corners:
                point = [
                    base_origin[0] + corner[0] * step[0],
                    base_origin[1] + corner[1] * step[1],
                    base_origin[2] + corner[2] * step[2],
                ]
                positions.append(point)
                normals.append(list(sdf_normal(sdf_tree, point, min(step) * 0.35)))
            indices.extend([base, base + 1, base + 2, base, base + 2, base + 3])
    return ProceduralMeshPayload(mcp_id=mcp_id, request_id=request_id, positions=positions, normals=normals, indices=indices)
